tests/integration/conftest.py

```diff
@@ -0,0 +1,147 @@
+#!/usr/bin/env python3
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+import logging
+import os
+from typing import Any, AsyncGenerator, Callable, Coroutine, Dict, Generator
+
+import pytest
+from lightkube import Client, KubeConfig
+from lightkube.core.exceptions import ApiError
+from lightkube.resources.apps_v1 import Deployment
+from playwright.async_api import async_playwright
+from playwright.async_api._generated import Browser, BrowserContext, BrowserType, Page
+from playwright.async_api._generated import Playwright as AsyncPlaywright
+from pytest_operator.plugin import OpsTest
+
+from integration.dex import apply_dex_resources, get_dex_manifest, get_dex_service_ip
+
+logger = logging.getLogger(__name__)
+KUBECONFIG = os.environ.get("TESTING_KUBECONFIG", "~/.kube/config")
+
+
+@pytest.fixture(scope="session")
+def client() -> Client:
+    return Client(config=KubeConfig.from_file(KUBECONFIG), field_manager="dex-test")
+
+
+@pytest.fixture(scope="module")
+def dex(ops_test: OpsTest, client: Client) -> Generator[str, None, None]:
+    # Use ops-lib-manifests?
+    try:
+        try:
+            client.get(Deployment, "dex", namespace="dex")
+        except ApiError:
+            logger.info("Deploying dex resources")
+            apply_dex_resources(client, restart=False)
+
+            # We need to patch the dex service to apply the metallb ip, we don't know that before
+            # the service is created
+            apply_dex_resources(client)
```

Comment:
I'm still a bit confused why we would `apply` a second time.

Reply:
I will update the comment.

The problem that we face is that when we first apply the dex manifests, we don't know which IP the dex service will get. We need to know the IP to configure the dex `issuer_url`, so once the service starts we call this function again to update the configuration and restart dex. If you can think of a cleaner approach for this, please let me know.